contrib/src/main/scala/akka/stream/contrib/FocusedFlow.scala

```diff
@@ -0,0 +1,241 @@
+package akka.stream.contrib
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.stream._
+import akka.stream.contrib.FocusedSource.IncompleteFocusedSourceShape
+import akka.stream.scaladsl.{BidiFlow, Flow, FlowOpsMat, GraphDSL, Keep, RunnableGraph, Sink, Source}
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+
+import scala.annotation.unchecked.uncheckedVariance
+import scala.collection.immutable
+
+object FocusedFlow {
+  implicit class FocusedFlowOps[Tag, Src, In, Mat](from: Flow[Src, (Tag, In), Mat]) {
+    def focus = new FocusedFlow(BidiFlow.fromFlowsMat(from, Flow[In])(Keep.left))
+  }
+
+  def main(args: Array[String]): Unit = {
+    implicit val system = ActorSystem()
+    implicit val materializer = ActorMaterializer()
+
+    val flow: Flow[(Int, String), (Int, String), NotUsed] = Flow[(Int, String)]
+      .focus
+        .map(_.toUpperCase)
+        .filterNot(_.startsWith("T"))
```

Comment:
This seems like a common problem whenever an API takes some flow as input without knowing anything on it.
I had similar problems in:
https://github.com/akka/akka-stream-contrib/pull/33
https://github.com/akka/akka-stream-contrib/pull/26
Input flow may buffer, expand or reduce elements, it may also complete early (and asymmetrically, e.g: only completing upstream, but not downstream...).
I don't think it's possible to ensure anything on the flow.
In mentioned PRs, I left most of these worries as the user responsibility,
and added this: https://github.com/akka/akka-stream-contrib/blob/master/contrib/src/main/resources/reference.conf
to ensure a better completion for the stream.
but everything else is left at the user responsibility, and documented with warning against misuse.
I wonder if it would be possible to add a more constrained type of flow,
something like `Unbuffered1To1RationFlow[I,O,M] <: Flow[I,O,M]` (with a better name ;)).
but that's probably out of scope here, and much harder to implement...